### PR TITLE
[kubeadm] use iteration instead of recursion in function

### PR DESCRIPTION
**Pull request by bruceauyeung**
**What this PR does / why we need it**:

before this PR, function `attemptToUpdateMasterRoleLabelsAndTaints` recursively call itself. there are some defeats in it:
1. potential stack overflow.
2. unnecessary extra  `json.Marshal` calls.
3. unnecessary extra `client.Nodes().List` calls.

this PR rewrite `attemptToUpdateMasterRoleLabelsAndTaints` function, use iterate instead of recursion.
so these 3 defeats the metioned above are gone.
Signed-off-by: bruceauyeung <ouyang.qinhua@zte.com.cn>

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36625)

### CODE REVIEW

**Comment by luxas** [comment]
@k8s-bot ok to test

**Comment by luxas** [comment]
LGTM if all tests pass

**Comment by luxas** [comment]
@k8s-bot cvm gce e2e test this

**Comment by luxas** [comment]
@k8s-bot gci gce e2e test this

**Comment by luxas** [comment]
@k8s-bot gci gke e2e test this

**Comment by luxas** [comment]
@k8s-bot kubemark e2e test this

**Comment by luxas** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by luxas** [comment]
@k8s-bot cvm gke e2e test this

**Comment by luxas** [comment]
@k8s-bot test this

**Comment by bruceauyeung** [comment]
@luxas should i rebase ?

File: cmd/kubeadm/app/master/apiclient.go [hunk 1]
```diff
@@ -177,21 +177,22 @@ func attemptToUpdateMasterRoleLabelsAndTaints(client *clientset.Clientset, sched
...
 	}
 
-	if _, err := client.Nodes().Update(n); err != nil {
-		if apierrs.IsConflict(err) {
-			fmt.Println("<master/apiclient> temporarily unable to update master node metadata due to conflict (will retry)")
-			time.Sleep(apiCallRetryInterval)
-			attemptToUpdateMasterRoleLabelsAndTaints(client, schedulable)
+	for {
+		if _, err := client.Nodes().Update(n); err != nil {
+			if apierrs.IsConflict(err) {
```
Comment (deads2k) [inline cmd/kubeadm/app/master/apiclient.go]: isn't this exactly what the `RetryOnConflict` is for? Also, how does ever succeed if you aren't changing `n`?

**Comment by deads2k** [comment]
This pull should probably be reverted.  It breaks the functionality.

File: cmd/kubeadm/app/master/apiclient.go [hunk 2]
```diff
@@ -177,21 +177,22 @@ func attemptToUpdateMasterRoleLabelsAndTaints(client *clientset.Clientset, sched
 		n.ObjectMeta.Annotations[api.TaintsAnnotationKey] = string(taintsAnnotation)
 	}
 
-	if _, err := client.Nodes().Update(n); err != nil {
-		if apierrs.IsConflict(err) {
-			fmt.Println("<master/apiclient> temporarily unable to update master node metadata due to conflict (will retry)")
-			time.Sleep(apiCallRetryInterval)
-			attemptToUpdateMasterRoleLabelsAndTaints(client, schedulable)
```
Comment (deads2k) [inline cmd/kubeadm/app/master/apiclient.go]: if you want this behavior, you should be building a patch.

**Comment by deads2k** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/37813 to revert this.

**Pull request merged**